nexus/db-queries/src/db/datastore/allowed_source_ips.rs

```diff
@@ -0,0 +1,250 @@
+
+use crate::authz;
+use crate::context::OpContext;
+use crate::db::error::public_error_from_diesel;
+use crate::db::error::ErrorHandler;
+use crate::db::fixed_data::allowed_source_ips::ALLOWED_SOURCE_IPS_ID;
+use crate::db::DbConnection;
+use async_bb8_diesel::AsyncRunQueryDsl;
+use diesel::ExpressionMethods;
+use diesel::QueryDsl;
+use diesel::SelectableHelper;
+use nexus_db_model::schema::allowed_source_ip;
+use nexus_db_model::AllowedSourceIp;
+use omicron_common::api::external::AllowedSourceIps;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::LookupType;
+use omicron_common::api::external::ResourceType;
+
+impl super::DataStore {
+    /// Fetch the list of allowed source IPs.
+    ///
+    /// This is currently effectively a singleton, since it is populated by RSS
+    /// and we only provide APIs in Nexus to update it, not create / delete.
+    pub async fn allowed_source_ips_view(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<AllowedSourceIp, Error> {
+        let conn = self.pool_connection_authorized(opctx).await?;
+        opctx.authorize(authz::Action::Read, &authz::FLEET).await?;
+        allowed_source_ip::dsl::allowed_source_ip
+            .find(ALLOWED_SOURCE_IPS_ID)
+            .first_async::<AllowedSourceIp>(&*conn)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel(
+                    e,
+                    ErrorHandler::NotFoundByLookup(
+                        ResourceType::AllowedSourceIps,
+                        LookupType::ById(ALLOWED_SOURCE_IPS_ID),
+                    ),
+                )
+            })
+    }
+
+    /// Upsert and return a list of allowed source IPs.
+    pub async fn allowed_source_ips_upsert(
+        &self,
+        opctx: &OpContext,
+        allowed_ips: AllowedSourceIps,
+    ) -> Result<AllowedSourceIp, Error> {
+        let conn = self.pool_connection_authorized(opctx).await?;
+        Self::allowed_source_ips_upsert_on_connection(
+            opctx,
+            &conn,
+            allowed_ips,
+        )
+        .await
+    }
+
+    pub(crate) async fn allowed_source_ips_upsert_on_connection(
+        opctx: &OpContext,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        allowed_ips: AllowedSourceIps,
+    ) -> Result<AllowedSourceIp, Error> {
+        use allowed_source_ip::dsl;
+        opctx.authorize(authz::Action::Modify, &authz::FLEET).await?;
+        let record = AllowedSourceIp::new(ALLOWED_SOURCE_IPS_ID, allowed_ips);
+        diesel::insert_into(dsl::allowed_source_ip)
+            .values(record.clone())
+            .returning(AllowedSourceIp::as_returning())
+            .on_conflict(dsl::id)
+            .do_update()
+            .set((
+                dsl::allowed_ips.eq(record.allowed_ips),
+                dsl::time_modified.eq(record.time_modified),
+                dsl::generation.eq(dsl::generation + 1),
```

Comment:
I opted to leave out the generation number for incoming requests for this PR. So what we'd do here is:

- fetch the row in the database, if it exists
- check the generation
- insert the next generation conditional on the generation number not having changed

I'm not sure I see the value of that. We're always going to simply insert the latest allowlist. It's one transaction already, so why split it into two when we don't do any other checks on the returned record? If we were going to plumb the generation back up to the _client_, then yeah, it would be important to confirm the generation is the same as the one the client supplied.

TBH, we could probably remove the generation column entirely. I kept it in there from an early version where it was passed to clients, but that's no longer the case. Maybe it's just not important anymore.

Reply:
I think if we're always going to accept the latest request, then yeah, drop the generation number entirely; having it makes it look like this is being managed like other generation-gated data, but it really isn't. I think if we wanted to use it, we wouldn't internally do the steps you outlined; we'd push that down to the operator, to avoid the dueling operator problem. They would need to specify the generation as part of the request to set a new list.

Reply:
Sounds good, I'll take it out. I think we want to encourage the operator to have their own, out-of-band mechanism for controlling updates to this list. Generations or ETags are not something we have anywhere else in the API, and it seems like it might not be the most appropriate solution.

Reply:
Done in d72f7dc